### PR TITLE
feat: macaroon based entity authentication

### PR DESCRIPTION
**Pull request by jackstar12**
Introduces the concept of `entities` to allow multiple users for a specific client. 

An entity can be created with the `CreateEntity`. A macaroon for the specific entity can be baked using `BakeMacaroon`. These macaroons include a caveat which restricts it to its specific entity. 

Swaps and wallets belong to a specific entity which will only be returned by requests with the appropiate macaroon while the admin can retrieve information for any entity.

The client will behave in standalone mode for any request that doesnt come from an admin.

### CODE REVIEW

File: boltzrpc/boltzrpc.proto [hunk 1]
```diff
@@ -138,7 +138,36 @@ service Boltz {
...
+    repeated EntityInfo entities = 1;
+}
+
+message EntityInfo {
+    int64 id = 1;
+    string name = 2;
+}
+
+message BakeMacaroonRequest {
+    string entity = 1;
```
Comment (michael1011) [inline boltzrpc/boltzrpc.proto:166]: I am torn
Why not use the `id` here? We expose it in `EntityInfo`, we might as well make use of it

File: boltzrpc/boltzrpc.proto [hunk 2]
```diff
@@ -138,7 +138,36 @@ service Boltz {
...
+}
+message BakeMacaroonResponse {
+    string admin_macaroon = 1;
+    string readonly_macaroon = 2;
```
Comment (michael1011) [inline boltzrpc/boltzrpc.proto:170]: How about being able to specify the constraints when baking it?

File: cln/cln.go [hunk 3]
```diff
@@ -95,7 +89,20 @@ func (c *Cln) Connect() error {
 }
 
 func (c *Cln) Name() string {
-	return string(serviceName)
+	return "CLN"
```
Comment (michael1011) [inline cln/cln.go:92]: Why not `string(serviceName)`?

File: cln/cln.go [hunk 4]
```diff
@@ -95,7 +89,20 @@ func (c *Cln) Connect() error {
...
+}
+
+func (c *Cln) GetWalletInfo() onchain.WalletInfo {
+	return c.walletInfo
+}
+
+func (c *Cln) SetId(id int64) {
+	c.walletInfo = onchain.WalletInfo{
```
Comment (michael1011) [inline cln/cln.go:100]: This way, `SetId` initializes the wallet. I don't like that naming. How about `SetupWallet`?

File: cln/cln.go [hunk 5]
```diff
@@ -33,7 +33,9 @@ type Cln struct {
 
 	Client protos.NodeClient
 
-	regtest bool
+	// id which is used to satisfy wallet interface
+	regtest    bool
```
Comment (michael1011) [inline cln/cln.go:37]: ?

File: cmd/boltzd/test/boltz.toml [hunk 6]
```diff
@@ -61,7 +61,7 @@ tlsCert = ""
 tlsKey = ""
 
 # Whether the macaroon authentication for the gRPC and REST interface should be disabled
-noMacaroons = true
+noMacaroons = false
```
Comment (michael1011) [inline cmd/boltzd/test/boltz.toml:64]: Let's not make that default

File: database/database.go [hunk 7]
```diff
@@ -57,7 +57,8 @@ CREATE TABLE swaps
     serviceFeePercent   REAL,
     onchainFee          INT,
     createdAt           INT,
-    wallet              VARCHAR
+    walletId            INT REFERENCES wallets(id) ON DELETE SET NULL,
+    entityId			      INT REFERENCES entities(id)
```
Comment (michael1011) [inline database/database.go:61]: Ident looks off

File: database/database.go [hunk 8]
```diff
@@ -87,7 +88,9 @@ CREATE TABLE reverseSwaps
     serviceFeePercent   REAL    DEFAULT 0,
     onchainFee          INT,
     createdAt           INT,
-    externalPay         BOOLEAN
+    externalPay         BOOLEAN,
+    walletId            INT REFERENCES wallets(id) ON DELETE SET NULL,
+    entityId			      INT REFERENCES entities(id)
```
Comment (michael1011) [inline database/database.go:93]: Same here

File: database/database.go [hunk 9]
```diff
@@ -96,14 +99,24 @@ CREATE TABLE autobudget
...
+    salt           VARCHAR,
+    entityId       INT REFERENCES entities(id),
+	
+    UNIQUE (name, entityId, nodePubkey),
+    UNIQUE (xpub, coreDescriptor, mnemonic, nodePubkey)
 );
+CREATE TABLE entities (
+    id INTEGER PRIMARY KEY AUTOINCREMENT,
+    name VARCHAR UNIQUE
+)
```
Comment (michael1011) [inline database/database.go:119]:
```suggestion
CREATE TABLE entities (
    id   INTEGER PRIMARY KEY AUTOINCREMENT,
    name VARCHAR UNIQUE
);
```

File: database/entity.go [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+	query := "INSERT INTO entities (name) VALUES (?) RETURNING id"
+	row := d.QueryRow(query, entity.Name)
+	return row.Scan(&entity.Id)
+}
+
+func (d *Database) GetEntity(id int64) (*Entity, error) {
+	d.lock.RLock()
+	defer d.lock.RUnlock()
+	query := "SELECT * FROM entities WHERE id = ?"
+	rows, err := d.Query(query, id)
```
Comment (michael1011) [inline database/entity.go:23]: Why not `QueryRow`? We only need one result anyways

File: database/entity.go [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+		return parseEntity(rows)
+	}
+	return nil, fmt.Errorf("could not find entity with id %d", id)
+}
+
+func (d *Database) GetEntityByName(name string) (*Entity, error) {
+	d.lock.RLock()
+	defer d.lock.RUnlock()
+	query := "SELECT * FROM entities WHERE name  = ?"
+	rows, err := d.Query(query, name)
```
Comment (michael1011) [inline database/entity.go:38]: Same here

File: database/entity.go [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+)
+
+type Entity struct {
+	Id   int64
+	Name string
+}
+
+func (d *Database) CreateEntity(entity *Entity) error {
+	query := "INSERT INTO entities (name) VALUES (?) RETURNING id"
+	row := d.QueryRow(query, entity.Name)
```
Comment (michael1011) [inline database/entity.go:15]: Does `QueryRow` execute `INSERT` statements?

File: onchain/onchain.go [hunk 13]
```diff
@@ -91,50 +105,37 @@ func (onchain *Onchain) GetCurrency(currency boltz.Currency) (*Currency, error)
...
-	var found []Wallet
+func (walletChecker *WalletChecker) Allowed(wallet Wallet) bool {
+	info := wallet.GetWalletInfo()
+	return wallet.Ready() &&
+		(walletChecker.Id == nil || info.Id == *walletChecker.Id) &&
+		(info.Currency == walletChecker.Currency || walletChecker.Currency == "") &&
+		(info.Name == walletChecker.Name || walletChecker.Name == "") &&
+		(!info.Readonly || walletChecker.Readonly) &&
+		(info.EntityId == walletChecker.EntityId || (info.EntityId != nil && walletChecker.EntityId != nil && *info.EntityId == *walletChecker.EntityId))
+}
```
Comment (michael1011) [inline onchain/onchain.go:116]: There has to be a nicer way to do this

File: onchain/onchain.go [hunk 14]
```diff
@@ -23,6 +23,22 @@ type Balance struct {
...
+	Currency boltz.Currency
+	Readonly bool
+	EntityId *int64
+}
+
+type WalletChecker struct {
+	Id       *int64
+	Name     string
+	Currency boltz.Currency
+	Readonly bool
```
Comment (michael1011) [inline onchain/onchain.go:38]: Not sure what `Readonly` means in that context? Also allow readonly wallet? Only allow readonly?

File: boltzrpc/boltzrpc.proto (hunk 1 above)
Comment (jackstar12) [reply boltzrpc/boltzrpc.proto:166]: More convenient to use in the cli. We could accept both

File: cmd/boltzd/test/boltz.toml (hunk 6 above)
Comment (jackstar12) [reply cmd/boltzd/test/boltz.toml:64]: Its just for the integration tests

File: database/entity.go (hunk 10 above)
Comment (jackstar12) [reply database/entity.go:23]: because `parseEntity` requires wants a `sql.Rows` type

File: database/entity.go (hunk 12 above)
Comment (jackstar12) [reply database/entity.go:15]: It works, so yes

File: onchain/onchain.go (hunk 14 above)
Comment (jackstar12) [reply onchain/onchain.go:38]: Its allow readonly. Can reword that

File: boltzrpc/boltzrpc.proto (hunk 2 above)
Comment (jackstar12) [reply boltzrpc/boltzrpc.proto:170]: Sounds good, will do an enum in the request

File: cln/cln.go (hunk 3 above)
Comment (jackstar12) [reply cln/cln.go:92]: accident

File: boltzrpc/boltzrpc.proto [hunk 15]
```diff
@@ -138,7 +138,64 @@ service Boltz {
...
+
+    /*
+    Get a specifiy entity.
+     */
+    rpc GetEntity(GetEntityRequest) returns (Entity);
+
+    /*
+    Bakes a new macaroon for a given entity with the specified permissions.
+    Any swap or wallet created with the returned macaroon will belong to this entity and can not be accessed by other entities.
+    The underlying lightning node of the daemon can not be used to pay or create invocies for swaps.
```
Comment (michael1011) [inline boltzrpc/boltzrpc.proto:161]:
```suggestion
    The lightning node connected to the daemon can not be used to pay or create invocies for swaps.
```

File: boltzrpc/boltzrpc.proto [hunk 16]
```diff
@@ -138,7 +138,64 @@ service Boltz {
...
+    READ = 0;
+    WRITE = 1;
+}
+
+message MacaroonPermissions {
+    MacaroonAction action = 2;
+}
+
+message BakeMacaroonRequest {
+    int64 entity_id = 1;
```
Comment (michael1011) [inline boltzrpc/boltzrpc.proto:194]:
```suggestion
    int64 entity_id = 1;
```

Could we make this optional to allow creating new macaroons for admin?

File: lnd/lnd.go [hunk 17]
```diff
@@ -58,22 +58,29 @@ type LND struct {
...
+	return "LND"
 }
 
-func (lnd *LND) Readonly() bool {
-	return false
+func (lnd *LND) SetupWallet(id int64) {
+	lnd.walletInfo = onchain.WalletInfo{
+		Id:       id,
+		Readonly: false,
+		Name:     "LND",
```
Comment (michael1011) [inline lnd/lnd.go:77]:
```suggestion
		Name:     lnd.Name(),
```

**Review by michael1011 — CHANGES_REQUESTED** [review]

**Pull request merged**